authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,156 @@
+
+    def __init__(self, ldap_url, search_properties=None):
+        self.ldap_url = ldap_url
+        self.search_props = search_properties
+        self.validate_search_props()
+
+    def validate_search_props(self):
+        if self.search_props:
+            valid = all(test in self.search_props for test in
+                        REQ_SEARCH_FIELDS)
+            if not valid:
+                err = 'The configured search_properties are missing required '\
+                      'search fields. Search properties must hold values for '\
+                      'the following fields: \'{req_fields}\', got '\
+                      '\'{props_fields}\''\
+                    .format(req_fields=REQ_SEARCH_FIELDS,
+                            props_fields=self.search_props.keys())
+                self._logger.warning(err)
+                raise RuntimeError(err)
+
+    # LDAP authentication can be done by subtree match or according to
+    # dn attribute.
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.search_props:
+            user_dn = self.get_dn_by_user_attribute(who,
+                                                    ldap_url=self.ldap_url,
+                                                    **self.search_props)
+        else:
+            user_dn = who
+        # attempt to bind using the matching user ID/DN and provided password.
+        self.validate_bind(user_dn, password)
+        return who
+
+    @classmethod
+    def handle_authentication_error(cls, err, conn):
+        conn.unbind()
+        cls._logger.warning(err)
+        raise AuthenticationException(err)
+
+    def validate_bind(self, who, password):
+            conn = self.create_bind_conn(self.ldap_url, who, password)
+            conn.unbind()
+
+    @classmethod
+    def get_dn_by_user_attribute(cls,
+                                 who,
+                                 ldap_url,
+                                 base_dn,
+                                 admin_user_id,
+                                 admin_password,
+                                 user_id_attribute=DEFAULT_AUTH_ATTR_NAME):
+
+        search_filter = '{0}={1}'.format(user_id_attribute, who)
+        try:
+            conn = cls.create_bind_conn(ldap_url,
+                                        admin_user_id,
+                                        admin_password)
+            # Search for matches under the base_dn subtree using the attribute
+            # search filter.
+            ldap_result_id = conn.search(base_dn,
+                                         ldap.SCOPE_SUBTREE,
+                                         search_filter,
+                                         [DN_ATTRIBUTE_NAME])
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
```

Comment:
seems like you should call `handle_authentication_error` and pass None as the `conn` argument.
In `handle_authentication_error` you should call `conn.unbind()` only if conn is not None.
That way any error handling actions added into `handle_authentication_error` will not be missing / duplicated here